File: src/heap.cpp
#include "heap.hpp"
#include "utils.hpp"

#include <ac-library/containers/persistent/binary_heap/binary_heap.hpp>
#include <cstdint>

namespace NACJSDiskStructs {
    using namespace NAC;

    Napi::Value HeapOpen(const Napi::CallbackInfo& info) {
        return Napi::Number::New(info.Env(), (uintptr_t)new TPersistentBinaryHeap(info[0].As<Napi::String>()));
    }

    Napi::Value HeapNew(const Napi::CallbackInfo& info) {
        return Napi::Number::New(info.Env(), (uintptr_t)new TPersistentBinaryHeap(info[0].As<Napi::String>(), info[1].As<Napi::Number>().Int64Value()));
    }

    Napi::Value HeapDelete(const Napi::CallbackInfo& info) {
        delete (TPersistentBinaryHeap*)AsPtr(info[0]);

        return info.Env().Undefined();
    }

    Napi::Value HeapInsert(const Napi::CallbackInfo& info) {
        auto value = info[1].As<Napi::Buffer<uint8_t>>();

        ((TPersistentBinaryHeap*)AsPtr(info[0]))->Insert((const char*)value.Data());

        return info.Env().Undefined();
    }

    Napi::Value HeapClose(const Napi::CallbackInfo& info) {
        ((TPersistentBinaryHeap*)AsPtr(info[0]))->Close();

        return info.Env().Undefined();
    }

    Napi::Value HeapRecordSize(const Napi::CallbackInfo& info) {
        return Napi::Number::New(info.Env(), ((TPersistentBinaryHeap*)AsPtr(info[0]))->RecordSize());
    }

    Napi::Value HeapGetAll(const Napi::CallbackInfo& info) {
        auto prefix = info[1].As<Napi::Buffer<uint8_t>>();

        return Napi::Number::New(info.Env(), (uintptr_t)new TPersistentBinaryHeap::TIterator(((TPersistentBinaryHeap*)AsPtr(info[0]))->GetAll(TBlob(prefix.Length(), (const char*)prefix.Data()))));
    }

    Napi::Value HeapIteratorDelete(const Napi::CallbackInfo& info) {
        delete (TPersistentBinaryHeap::TIterator*)AsPtr(info[0]);

        return info.Env().Undefined();
    }

    Napi::Value HeapIteratorNext(const Napi::CallbackInfo& info) {
        TBlob data;

        if (((TPersistentBinaryHeap::TIterator*)AsPtr(info[0]))->Next(data)) {
            return Napi::Buffer<uint8_t>::Copy(info.Env(), (uint8_t*)data.Data(), data.Size());
        }

        return info.Env().Undefined();
    }
}

File: wrappers/heap.js
const LIB = require('bindings')('ac_diskstructs');

function HeapIterator(ptr) {
    const self = this;

    self.delete = function() {
        LIB.HeapIteratorDelete(ptr);

        self.delete = function() {};
    };

    self[Symbol.iterator] = function*() {
        for (let item; (item = LIB.HeapIteratorNext(ptr)) !== undefined;) {
            yield item;
        }

        self.delete();
    };
}

module.exports = function(path, {recordSize} = {}) {
    const self = this;
    const ptr = (recordSize ? LIB.HeapNew(path, recordSize) : LIB.HeapOpen(path));

    self.recordSize = function() {
        return LIB.HeapRecordSize(ptr);
    };

    // Free memory
    self.delete = function() {
        LIB.HeapDelete(ptr);

        self.delete = function() {};
    };

    // Finish write operations
    self.close = function() {
        LIB.HeapClose(ptr);

        self.close = function() {};
    };

    self.get = function(prefix) {
        const it = self.getAll(prefix);

        for (let rv of it) {
            it.delete();
            return rv;
        }

        return undefined;
    };

    self.insert = function(value) {
        LIB.HeapInsert(ptr, (Buffer.isBuffer(value) ? value : Buffer.from(value)));
    };

    self.getAll = function(prefix) {
        return new HeapIterator(LIB.HeapGetAll(ptr, (Buffer.isBuffer(prefix) ? prefix : Buffer.from(prefix))));
    };
};
